designs/2024-repo-ecosystem-plugin-tests/README.md

```diff
@@ -0,0 +1,183 @@
+- Repo: eslint/eslint
+- Start Date: 2024-11-25
+- RFC PR: <https://github.com/eslint/rfcs/pull/127>
+- Authors: [Josh Goldberg](https://github.com/JoshuaKGoldberg)
+
+# Introduce ecosystem tests for popular plugins
+
+## Summary
+
+Adding an CI job to the `eslint/eslint` repository that checks changes against `@eslint/*` plugins as well as a small selection of third-party plugins.
+
+## Motivation
+
+Changes in ESLint occasionally break downstream plugins in unexpected ways.
+Those changes might be unintentional breaking changes, or even non-breaking changes that happen to touch edge case behaviors relied on by plugins.
+
+[Bug: Error while loading rule '@typescript-eslint/no-unused-expressions'](https://github.com/eslint/eslint/issues/19134) reports an example change in ESLint that caused downstream breakages in third-party plugins.
+At least two popular plugins -[`eslint-plugin-unicorn`](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/2496) and [`typescript-eslint`](https://github.com/typescript-eslint/typescript-eslint/issues/10338)- were broken by that change.
+
+The plugins broke because they were relying on non-public implementation details of ESLint rules per [Docs: Formalize recommendation against plugins calling to rules via use-at-your-own-risk](https://github.com/eslint/eslint/issues/19169).
+ESLint core's [`eslint-config-eslint`](https://github.com/eslint/eslint/tree/main/packages/eslint-config-eslint) does not use all rules of downstream plugins and is not always up-to-date with their latest versions, so its internal usage of plugins is not sufficient to flag all high visibility compatibility issues.
+When the root cause is a bug in the downstream plugins, an "early warning" system would help them fix their issues before the incompatible changes to ESLint are published.
+
+## Detailed Design
+
+This RFC proposes creating a small list of popular third-party plugins that will be tested as part of ESLint's CI.
+Each plugin will have a `test:eslint-compat` script in their `package.json` that runs lint rule tests.
+
+See [Plugin Selection](#plugin-selection) below for specifics on which plugins will be included.
+
+> ⚠️ Plugins are currently being asked for feedback on the `test:eslint-compat` script.
+
+### CI Job
+
+The new CI job will, for each plugin:
+
+1. Clone the plugin into a directory named `test/ecosystem/${plugin}`
+2. Run the plugin's package installation and build commands with [ni](https://github.com/antfu-collective/ni)
+3. Run the plugin's `test:eslint-compat` script with [ni](https://github.com/antfu-collective/ni)
```

Comment:
How will we get the script to use the version of eslint from the branch we are testing instead of the installed one?

Reply:
Ah, great point. We'll have to `npm link` / an equivalent. Which is not supportable in ni per https://github.com/antfu-collective/ni/issues/85. Added.

We could always spin out a dedicated standalone script for the single use case of linking a child directory to a parent...